### PR TITLE
Feat/gba 16

### PR DESCRIPTION
**Pull request by jesusdmedinac**
Clean Architecture implemented.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -3,13 +3,14 @@ apply plugin: 'kotlin-android'
...
 apply plugin: 'kotlin-kapt'
+apply plugin: "androidx.navigation.safeargs.kotlin"
 
 android {
     compileSdkVersion 29
     buildToolsVersion "29.0.2"
     defaultConfig {
         applicationId "com.mupper.gobus"
-        minSdkVersion 21
+        minSdkVersion 24
```
Comment (proyectox123) [inline app/build.gradle:13]: Did the minimum sdk version change for business or technical decision?

File: app/src/main/java/com/mupper/gobus/commons/extension/Activity.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+@Suppress("UNCHECKED_CAST")
+inline fun <reified T : ViewModel> FragmentActivity.getViewModel(crossinline factory: () -> T): T {
+
+    val vmFactory = object : ViewModelProvider.Factory {
+        override fun <U : ViewModel> create(modelClass: Class<U>): U = factory() as U
+    }
+
+    return ViewModelProviders.of(this, vmFactory)[T::class.java]
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/commons/extension/Activity.kt:21]: I saw that ViewModelProviders is deprecated (https://developer.android.com/reference/androidx/lifecycle/ViewModelProviders).

I implemented in this way:

return ViewModelProvider(this.viewModelStore, vmFactory)[T::class.java]

File: app/src/main/java/com/mupper/gobus/data/database/GobusDatabase.kt [hunk 3]
```diff
@@ -1,24 +1,26 @@
...
 
 /**
  * Created by jesus.medina on 12/2019.
- * Insulet Corporation
- * Andromeda
+ * Mupper
  */
 @Database(
-    entities = [Traveler::class],
+    entities = [Traveler::class, Bus::class],
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/data/database/GobusDatabase.kt:18]: Properly, modifications in database has to handle migration (https://developer.android.com/training/data-storage/room/migrating-db-versions).

File: app/src/main/java/com/mupper/gobus/commons/extension/Context.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package com.mupper.gobus.commons.extension
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/commons/extension/Context.kt:1]: Great! I think I'm going to split my extensions too :P

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        val colorPicker = SpectrumDialog.Builder(requireActivity())
+        colorPicker.setColors(R.array.md_colors)
+        color?.let { colorPicker.setSelectedColor(it) }
+        colorPicker.setOnColorSelectedListener { _, colorInt ->
+            busViewModel.onColorPicked(colorInt)
+        }
+        colorPicker.build().show(childFragmentManager, "¿De qué color es el camión?")
+    }
+
+    override fun onSelected() {
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt:75]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    override fun verifyStep(): VerificationError? = null
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt:78]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+    override fun onError(error: VerificationError) {
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt:80]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusCapacityFragment.kt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                )
+            }
+
+        binding?.apply {
+            bus = busViewModel
+            lifecycleOwner = this@NewBusCapacityFragment
+        }
+    }
+
+    override fun onSelected() {}
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusCapacityFragment.kt:60]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusCapacityFragment.kt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    override fun verifyStep(): VerificationError? = null
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusCapacityFragment.kt:62]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusCapacityFragment.kt [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+    override fun onError(error: VerificationError) {}
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusCapacityFragment.kt:64]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt (hunk 5 above)
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt:75]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt (hunk 6 above)
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt:78]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt (hunk 7 above)
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt:80]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathNameFragment.kt [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+                )
+            }
+
+        binding?.apply {
+            bus = busViewModel
+            lifecycleOwner = this@NewBusPathNameFragment
+        }
+    }
+
+    override fun onSelected() {}
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathNameFragment.kt:58]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathNameFragment.kt [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+    override fun verifyStep(): VerificationError? = null
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathNameFragment.kt:60]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathNameFragment.kt [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+    override fun onError(error: VerificationError) {}
```
Comment (proyectox123) [inline app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathNameFragment.kt:62]: Related to SOLID (Interface Segregation Principle), the client shouldn't be forced to depend on methods it doesn't use.

File: commons/src/main/res/values/colors.xml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <color name="white">#FFFFFF</color>
```
Comment (proyectox123) [inline commons/src/main/res/values/colors.xml:3]: Wouldn't it be convenient to use android color white resource?

File: features/src/main/java/com/mupper/features/bus/GetTravelingBus.kt [hunk 15]
```diff
@@ -0,0 +1,19 @@
+package com.mupper.features.bus
+
+import com.mupper.data.source.firestore.BusRemoteDataSource
+import com.mupper.data.source.room.BusLocalDataSource
+import com.mupper.domain.relations.BusWithTravelers
+
+class GetTravelingBus(
```
Comment (proyectox123) [inline features/src/main/java/com/mupper/features/bus/GetTravelingBus.kt:7]: Is this a repository? In that case, wouldn't it be convenient to use the suffix "Repository"?

File: app/build.gradle (hunk 1 above)
Comment (jesusdmedinac) [reply app/build.gradle:13]: I didn't remember, but I think it was technical decision.

File: app/src/main/java/com/mupper/gobus/commons/extension/Activity.kt (hunk 2 above)
Comment (jesusdmedinac) [reply app/src/main/java/com/mupper/gobus/commons/extension/Activity.kt:21]: Great!, I fixed it.

File: app/src/main/java/com/mupper/gobus/data/database/GobusDatabase.kt (hunk 3 above)
Comment (jesusdmedinac) [reply app/src/main/java/com/mupper/gobus/data/database/GobusDatabase.kt:18]: I think that is a good practice, but I'm not very sure If is a good idea implement it in an app that is not in a production environment.

File: app/src/main/java/com/mupper/gobus/commons/extension/Context.kt (hunk 4 above)
Comment (jesusdmedinac) [reply app/src/main/java/com/mupper/gobus/commons/extension/Context.kt:1]: I saw that you implement your data mappers similarly.

File: app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt (hunk 5 above)
Comment (jesusdmedinac) [reply app/src/main/java/com/mupper/gobus/ui/bus/steps/NewBusPathColorFragment.kt:75]: Those methods are from external library. 
I will implement their functionalities later.

File: commons/src/main/res/values/colors.xml (hunk 14 above)
Comment (jesusdmedinac) [reply commons/src/main/res/values/colors.xml:3]: Yes, I didn't remember that Android has internally those colors.

File: features/src/main/java/com/mupper/features/bus/GetTravelingBus.kt (hunk 15 above)
Comment (jesusdmedinac) [reply features/src/main/java/com/mupper/features/bus/GetTravelingBus.kt:7]: That is a feature, in others words, is a use case, so, I think in practice is the same think a Repository and a Use case, it depends on your methodology.

What I wanted to implement is a Use Case

**Pull request merged**